### DISCUSSION TITLE
How to set mappings initially

### DISCUSSION THREAD

**Discussion by masterplanx** [jertel/elastalert2#527] [Q&A] [answered]
Hi team! I am trying to build a dashboard, to get the historical alerts, I tried to use for example match_body.error.message field, but from kibana I received the message "Unknown field", so looks like is not indexing and is not possible to make visualizations.
**For example:**
![image](https://user-images.githubusercontent.com/12055239/138926233-bbe2c740-2664-464a-934f-2f920e394139.png)
I saw here https://github.com/jertel/elastalert2/blob/master/elastalert/es_mappings/6/elastalert.json#L31 that the match_body is false

```
    "match_body": {
      "enabled": "false",
      "type": "object"
    },

```
So I guess we have to change that mapping to enable it, is that correct? If it is, there is a proper way to initialize the index?
I look at the Enhancement module but am not sure how to use it, we are using the official helm chart version 2.2.2.
can you help me with that?

Thanks and regards!

**Comment by jertel** [comment]
You can adjust the es_mappings via the Elasticsearch API. Ex: https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-put-mapping.html

Since you're using the Helm chart that's probably the quickest way. If you were running ElastAlert manually then you could override the es_mappings and have it recreate the indices.

**Comment by masterplanx** [comment]
Hi @jertel thanks for the answer, yes  I tried to update the mapping but I can't , I ran this 

```
PUT elastalert/_mapping
{
    "properties": {
      "match_body": {
        "type": "object",
        "enabled": true
    }
 }
}
```
But I got this result message:

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "mapper_exception",
        "reason" : "the [enabled] parameter can't be updated for the object mapping [match_body]"
      }
    ],
    "type" : "mapper_exception",
    "reason" : "the [enabled] parameter can't be updated for the object mapping [match_body]"
  },
  "status" : 500
}
```
looks like is not possible to update it (I am using 7.15.0 kibana version)

Any idea about this?

**Comment by jertel** [comment, marked as answer]
It appears that is a special setting that Elastic deliberately prevents changing. See https://github.com/elastic/elasticsearch/issues/33566

So in that case if you must create visualizations on those fields then you'll need to delete the indices that were created by the Helm chart deployment, and instead pre-create those indices with a custom es_mapping file prior to launching the helm chart. You could do this by running the ElastAlert 2 Docker container and mounting a customized es_mapping/... file into the container. On the first run it will detect that it needs to create the indices. Then once created, stop and remove the container, and switch to the Helm chart.

**Comment by masterplanx** [comment]
Thanks, @jertel , Do you think is a good idea to modify the es_mapping JSON files and  rebuild the docker image? I mean using the Dockerfile that is into the master branch. Wdyt?

Regards!

**Comment by jertel** [comment]
See https://github.com/Yelp/elastalert/issues/2093

**Reply by masterplanx** [reply]
@jertel yes I saw it but I don't understand how works that module with a helm chart, do you have some docs about it?

**Reply by jertel** [reply]
If you're talking about the enhancement you can build your own docker container and adjust the Helm values.yaml to use your image instead of the `jerte/elastalert2` image. Your image can have either the custom enhancement, or the custom es_mappings file.

But I primarily pasted it because you asked about modifying the public Dockerfile image and that post has some ideas on potential implications on making that change permanent for all ElastAlert 2 users.

**Reply by masterplanx** [reply]
good, I got it, I will take a look

**Comment by msteenhu** [comment]
We have the same problem.. I wonder if the 'disabled' match_body field could not be a setting? I will patch the source to remove that line since it can account for a lot of head scratching if we would delete the index and let ElastAlert2 rebuild it (with the problematic mapping).

If I understand it correctly, the disable is there since ElastAlert does not control the data in there and there can be duplicate fields with incompatible data types. But in our case we will make sure the data of different indices that will end up in 'match_body' will be compatible.

If I understand it correctly, and there is interest, I might be able to spend a bit time in an effort to make this configurable. So guess we need some jinja templating in the mappings or something like that to accomplish this?

**Reply by jertel** [reply]
`create_index.py` loads the JSON mappings into dict objects, so there's no need for adding Jinja complexity. You would simply override the enabled key on the dict after loading it from JSON.

This is a fairly old discussion. If you need to have more in-depth discussion on this before submitting a PR please start a new discussion.
